### DISCUSSION TITLE
JSONB are generated as string types with openapi-typescript 3.2.0

### DISCUSSION THREAD

**Discussion by burggraf** [supabase/supabase#984] [Questions]
I'm generating my typescript types as shown in the documentation, using:
```npx openapi-typescript <my-url> --output ./models/Supabase.ts```

This works great, except for my JSONB columns, which get translated to string types in my definitions.  These should be objects, or any. 

** context
I'm trying to store multiple addresses for a client in a JSONB object as an array of address objects.
```[ { streetaddress: "123 main st", city: "anytown", state: "GA" }, { streetaddress: "456 elm st", city: "anothertown", state: "GA"} ]```

But when I use the generated typescript interface, it's forcing the address field to a string.

**Comment by burggraf** [comment]
Ok, this gets even more complicated:

1. In the database, JSONB gets stored as a string, at least with the insert or upsert syntax.
2. In Typescript we want to deal with it as an object (a JSON object, to be precise).
3. So after reading data coming in from the database, we get strings, which we need to convert to JSON with JSON.parse.  Now we can deal with that data in Typescript.
4. Before we save this Typescript data back to the database, we need to convert it back to a string with JSON.stringify.

Is this the right way to handle JSONB data from Postgresql or am I missing something here?

**Comment by steve-chavez** [comment]
Hey @burggraf,

Yes, right now the openapi output defaults to string for a jsonb type.

According to openapi [object type](https://swagger.io/docs/specification/data-models/data-types/#object), we'd need to specify the exact properties of the object - which we can't for jsonb since it's loosely defined.

Hm, how would you suggest we map the jsonb type to a typescript type?

**Reply by burggraf** [reply]
@steve-chavez can you tell me where to go to make this change?  Maybe I can create a PR, if I knew where to go.

**Reply by steve-chavez** [reply]
The OpenAPI output is done by PostgREST, the change would have to be in these [lines](https://github.com/PostgREST/postgrest/blob/9f074cecce66ca4efa463f7bac5d05e9e155e133/src/PostgREST/OpenAPI.hs#L52-L63).

Unfortunately, right now it's not easy to extend the OpenAPI output because we use an outdated OpenAPI Haskell library([doesn't support OpenAPI v3](https://github.com/GetShopTV/swagger2/issues/105), among other things). However we have a pending enhancement that would make the OpenAPI output customizable with SQL.

**Reply by burggraf** [reply]
I'm a little confused about how this all fits together but in my output, the JSONB fields look like this:
```
                "email": {
                    "format": "jsonb",
                    "type": "string"
                },
```
So it seems like it should be relatively trivial to change the type to "any" for items that have format "jsonb".  A the very least I could write a post-processor on the output to generate models the way I want them, no?

**Reply by burggraf** [reply]
To take this a little further, I could even store a more complex typescript object type along with the column in the database inside the description field.  For example:
```
comment on column public.orgs.address is 'typescript:Address'
```
Now, my post-processor could parse the output from "npx openapi-typescript https://<myurl>" and look for the "typescript:" comments and use those to create more descriptive models, like:
```
export interface orgs: {
   id: string;
   address: Address;
}
```
where I have Address defined:
```
export interface Address {
  type?: string;
  pref?: boolean;
  formatted?: string;
  name?: string; // location name
  streetaddress?: string;
  locality?: string; // city
  region?: string; // state
  postalcode?: string; // zip
  country: string;
}
```
This would give me much better Typescript typing.

**Reply by burggraf** [reply]
Ok, I present to you: **supabase-to-typescript**

https://github.com/burggraf/supabase-to-typescript

**Comment by burggraf** [comment]
Map it to “any”.  That solves the problem, at least from a Typscript-throwing-errors point of view.  That would allow the field to be a string when you need to write it to Postgresql and an object when you want to manipulate it with Typescript.

I went ahead and wrote a couple functions in my code:  SqlToJSON to convert the strings into JSON objects with JSON.parse, and JSONtoSql to convert them back using JSON.stringify when writing back to the database.  Works just fine.

**Comment by mikebarkmin** [comment]
Hi. I come across the same issue. I resolved it by using this little helper script:

```javascript
import fs from "fs";
import openapiTS from "openapi-typescript";
import path from "path";

const updateTypes = async () => {
  const openapiUrl = `${process.env.NEXT_PUBLIC_SUPABASE_URL}/rest/v1/?apikey=${process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY}`;
  const output = await openapiTS(openapiUrl, {
    formatter: (node) => {
      if (node.format === "jsonb") {
        return "any";
      }
    },
  });

  fs.writeFileSync(path.join("src", "types", "supabase.ts"), output);
};

updateTypes();
``` 

It basically does what you have suggested. It uses any as a type, if the format is jsonb.

**Reply by aquaductape** [reply]
How did you run this script? I keep getting [ERR_REQUIRE_ESM]. https://github.com/drwpow/openapi-typescript/issues/928

**Comment by baptisteArno** [comment]
Thank you @mikebarkmin 

Here is my two cents for less permissive / more robust types:

```ts
import fs from "fs";
import openapiTS from "openapi-typescript";
import path from "path";

const updateTypes = async () => {
  const openapiUrl = `${process.env.NEXT_PUBLIC_SUPABASE_URL}/rest/v1/?apikey=${process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY}`;
  const output = await openapiTS(openapiUrl, {
    formatter: (node) => {
      if (node.format === "jsonb") {
        return "unknown";
      }
      if (node.format === "ARRAY") {
        return `${node.type}[]`;
      }
      if (node.format === "timestamp with time zone") {
        return `string | Date`;
      }
    },
  });

  fs.writeFileSync(path.join("src", "models", "v2", "dto.ts"), output);
};

updateTypes();

```

- It sets the type as `unknown` instead of `any` which is more accurate since the data format can't be trusted so you'll have to validate the format when deserializing the data.

- It handles arrays.
- It sets the "timestamp with time zone" format as `string | Date` type. Since supabase is OK when we send a `Date` object in a payload

**Reply by aquaductape** [reply]
How did you run this script? I keep getting [ERR_REQUIRE_ESM]. https://github.com/drwpow/openapi-typescript/issues/928
